### ISSUE TITLE
serviceMonitorNamespaceSelector doesn't work 

### ISSUE THREAD

**Issue by emas80** [prometheus-operator/prometheus-operator#3450]
We have a problem on two different setup (which we call prod and prod-infra) with the namespace label selector: the service monitors are not found. 
A third setup (stage), which to us looks EXACTLY like the other two, works perfectly well.

**What happened?**
This is the prometheus definition of the prod-infra, which DOES NOT work
```
apiVersion: monitoring.coreos.com/v1
kind: Prometheus
metadata:
  labels:
    app: prometheus-istio
  name: prometheus-istio
  namespace: istio-system
spec:
  image: quay.io/prometheus/prometheus
  enableAdminAPI: true
  replicas: 2
  ruleSelector:
    matchLabels:
      app: prometheus-istio
      release: prometheus-istio
  secrets: [ istio.prometheus ]
  serviceAccountName: prometheus
  serviceMonitorNamespaceSelector:
    matchLabels:
      prometheus/serviceMonitor: "true"
  serviceMonitorSelector: {}
  version: v2.19.1
```

This one instead works:
```
apiVersion: monitoring.coreos.com/v1
kind: Prometheus
metadata:
  labels:
    app: prometheus-istio
  name: prometheus-istio
  namespace: istio-system
spec:
  image: quay.io/prometheus/prometheus
  ruleNamespaceSelector: {}
  ruleSelector:
    matchLabels:
      app: prometheus-istio
      release: prometheus-istio
  secrets: [ istio.prometheus ]
  serviceAccountName: prometheus
  serviceMonitorNamespaceSelector:
    matchLabels:
      prometheus/serviceMonitor: "true"
  serviceMonitorSelector: {}
  version: v2.19.1
```
Yes, they look like the same because THEY ARE  the same. The only differences are in the resources section because one (the working one) is our stage environment, while the other two are the prod environments.

The namespace is this one - working one (stage):

```
apiVersion: v1
kind: Namespace
metadata:
  creationTimestamp: "2019-11-06T12:02:01Z"
  labels:
    prometheus/serviceMonitor: "true"
  name: istio-system
  resourceVersion: "337210798"
  selfLink: /api/v1/namespaces/istio-system
  uid: 3debf601-008d-11ea-83b5-42010a07000a
spec:
  finalizers:
  - kubernetes
status:
  phase: Active

```

Not working one (prod-infra):
```
apiVersion: v1
kind: Namespace
metadata:
  creationTimestamp: "2019-11-06T12:02:01Z"
  labels:
    prometheus/serviceMonitor: "true"
  name: istio-system
  resourceVersion: "337210798"
  selfLink: /api/v1/namespaces/istio-system
  uid: 3debf601-008d-11ea-83b5-42010a07000a
spec:
  finalizers:
  - kubernetes
status:
  phase: Active

```
If we instead of setting
```
  serviceMonitorNamespaceSelector:
    matchLabels:
      prometheus/serviceMonitor: "true"
```
we put

```
  serviceMonitorNamespaceSelector: {}
```
then also the others work and the service monitors - ALL the service monitors of the entire cluster - are discovered.

**Environment**

* Prometheus Operator version:
0.40.0

* Kubernetes version information:
GKE, 1.15.12-gke.2

**Anything else we need to know?**:
We have tried what we think are all the possible combinations of labelling the namespace, reinstalling prometheus operator or reinstalling this prometheus, and we still have absolutely no clue about why the service monitors are not discovered.

Everything else looks like the same on the three clusters: we have a git repo containing all the files, we do not apply random yaml from our machines or do any editing directly on the cluster.

You will see Prometheus running inside of the `istio-system` namespace, we have 
```
  namespaces:
    releaseNamespace: true
    additional:
    - istio-system
```
correctly set on all of the three prometheus operators.

I don't attach the logs because there are no logs.

Do you have any clue on what could possibly be wrong here?

**Comment by feliperoveran** [comment]
according to the docs, you need to use `matchNames`

https://github.com/prometheus-operator/prometheus-operator/blob/master/Documentation/api.md#namespaceselector

AFAIK it doesn't support selecting namespaces by labels, but I might be wrong :)

**Comment by emas80** [comment]
@feliperoveran 
According to this 

https://github.com/prometheus-operator/prometheus-operator/blob/master/Documentation/api.md#prometheusspec 

what we've done is correct.

We have upgraded prometheus to 2.19.3 and the operator helm chart to the latest version 9.3.1 (we did not update helm operator has the latest version is not supported by the current chart because of the `probes` resource).

This triggered a complete re-release of the operator and the managed prometheus and for some reasons it fixed our issue: the service monitors are discovered.

We still think there is an issue not specifically related to the versions of prometheus or of the helm chart - upgrading the versions triggered a complete "refresh" of some configurations that in turns made prometheus to have the updated configuration.
